src/server/debug/server/server.go

```diff
@@ -208,73 +143,418 @@ func (s *debugServer) appLogs(ctx context.Context, tw *tar.Writer) (retErr error
+						if err != nil {
+							return errors.Wrap(err, "open file")
+						}
+						defer func() {
+							if err := f.Close(); err != nil {
+								errors.JoinInto(&retErr, err)
+							}
+						}()
+						_, err = io.Copy(w, f)
+						return err
+					}
+					c, err := workerserver.NewClient(pod.Ip)
+					if err != nil {
+						return errors.Wrapf(err, "create worker client for IP %q", pod.Ip)
+					}
+					binaryC, err := c.Binary(ctx, &debug.BinaryRequest{})
+					if err != nil {
+						return errors.Wrap(err, "binary client")
+					}
+					r := grpcutil.NewStreamingBytesReader(binaryC, nil)
+					defer func() {
+						if err := r.Close(); err != nil {
+							errors.JoinInto(&retErr, errors.Wrapf(err, "close binary writer for pod %q", pod.Name))
+						}
+					}()
+					_, err = io.Copy(w, r)
+					return err
+				}); err != nil {
+					errors.JoinInto(&errs, errors.Wrapf(err, "close binary writer for pod %q", pod.Name))
+				}
+			}
+			rp(ctx)
+		}
+		return errs
+	}
+}
+
+func appDir(app *debug.App) string {
+	if app.Pipeline != nil {
+		return filepath.Join("pipelines", app.Pipeline.Project, app.Pipeline.Name)
+	}
+	return app.Name
+}
+
+type dumpContentServer struct {
+	server debug.Debug_DumpV2Server
+}
+
+func (dcs *dumpContentServer) Send(bytesValue *types.BytesValue) error {
+	return dcs.server.Send(&debug.DumpChunk{Chunk: &debug.DumpChunk_Content{Content: &debug.DumpContent{Content: bytesValue.Value}}})
+}
+
+type incProgressFunc func(context.Context)
+
+// TODO should this be sending in a goro?
+func recordProgress(server debug.Debug_DumpV2Server, task string, total int) incProgressFunc {
+	inc := 0
+	return func(ctx context.Context) {
+		inc++
+		if err := server.Send(
+			&debug.DumpChunk{
+				Chunk: &debug.DumpChunk_Progress{
+					Progress: &debug.DumpProgress{
+						Task:     task,
+						Progress: int64(inc),
+						Total:    int64(total),
+					},
+				},
+			},
+		); err != nil {
+			log.Error(ctx, fmt.Sprintf("error sending progress for task %q", task), zap.Error(err))
+		}
+	}
+}
+
+// TODO: sharpen timeouts
+func (s *debugServer) dump(c *client.APIClient, server debug.Debug_DumpV2Server, tasks []taskFunc) (retErr error) {
+	ctx := c.Ctx() // this context has authorization credentials we need
+	dumpRoot := filepath.Join(os.TempDir(), uuid.NewWithoutDashes())
+	if err := os.Mkdir(dumpRoot, 0744); err != nil {
+		return errors.Wrap(err, "create dump root directory")
+	}
+	defer func() {
+		errors.JoinInto(&retErr, os.RemoveAll(dumpRoot))
+	}()
+	timeout := 25 * time.Minute // If no client-side deadline set, use 25 minutes.
+	if deadline, ok := server.Context().Deadline(); ok {
+		d := time.Until(deadline)
+		// Time out our own operations at ~90% of the client-set deadline. (22 minutes of
+		// server-side processing for every 25 minutes of debug dumping.)
+		timeout = time.Duration(22) * d / time.Duration(25)
+		var cf context.CancelFunc
+		ctx, cf = context.WithTimeout(ctx, timeout)
+		defer cf()
+	}
```

Comment:
I think the ctx.WithTimeout needs to be dedented from the deadline ok block.  If the deadline isn't OK, then we want to timeout at 25 minutes.